apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -42,14 +55,93 @@ func (host *BareMetalHost) validateChanges(old *BareMetalHost) []error {
 	return errs
 }
 
-func validateRAID(r *RAIDConfig) error {
-	if r == nil {
+func validateRAID(s BareMetalHostSpec) error {
+	if s.RAID == nil {
 		return nil
 	}
 
-	if len(r.HardwareRAIDVolumes) > 0 && len(r.SoftwareRAIDVolumes) > 0 {
+	if len(s.RAID.HardwareRAIDVolumes) > 0 && len(s.RAID.SoftwareRAIDVolumes) > 0 {
 		return fmt.Errorf("hardwareRAIDVolumes and softwareRAIDVolumes can not be set at the same time")
 	}
 
+	bmcType, err := getBMCType(s.BMC.Address)
+	if err != nil {
+		return err
+	}
+
+	switch bmcType {
+	case "ilo5", "idrac", "irmc":
+		return nil
+	default:
+		return fmt.Errorf("currently the 'raid' field is only supported by ilo5/idrac/irmc")
+	}
+}
+
+func getBMCType(address string) (string, error) {
+	strs := strings.Split(address, "://")
+	if len(strs) != 2 {
+		return "", fmt.Errorf("the format of bmc.address: %s is wrong", address)
+	}
+
+	return strings.Split(strs[0], "+http")[0], nil
+}
+
+func validateBMC(b BMCDetails) error {
+	if b.Address == "" {
+		return nil
+	}
+
+	avaliableType := []string{
+		"ipmi", "idrac", "idrac-virtualmedia", "idrac-redfish",
+		"irmc", "ibmc", "ilo4", "ilo5", "ilo5-redfish",
+		"redfish", "redfish-virtualmedia", "libvirt",
```

Comment:
We absolutely should not hardcode a list of drivers in the API, even if we don't actually believe that we'll ever have non-Ironic provisioners.